chia/full_node/mempool_manager.py

```diff
@@ -80,6 +93,84 @@ def validate_clvm_and_signature(
     return None, bytes(result), new_cache_entries
 
 
+def run_for_cost_and_additions(
+    coin_id: bytes32, puzzle_reveal: SerializedProgram, solution: SerializedProgram, maximum_cost: uint64
+) -> Optional[Tuple[uint64, List[Coin]]]:
+    try:
+        cost, conditions = puzzle_reveal.run_with_cost(maximum_cost, solution)
+        created_coins = []
+        for condition in conditions.as_python():
+            if condition[0] == ConditionOpcode.CREATE_COIN:
+                puzzle_hash = bytes32(condition[1])
+                amount = uint64(int.from_bytes(condition[2], "big"))
+                created_coins.append(Coin(coin_id, puzzle_hash, amount))
+        return (uint64(cost), created_coins)
+    except Exception:
+        return None
+
+
+def check_item_for_dedup(
+    item: MempoolItem, dedup_coin_spends: Dict[bytes32, DedupCoinSpend]
+) -> Optional[Tuple[List[CoinSpend], uint64, Dict[bytes32, DedupCoinSpend], Set[Coin]]]:
+    consider_this_item = True
+    cost_saving = uint64(0)
+    spends_to_dedup: List[CoinSpend] = []
+    dedup_additions: Set[Coin] = set()
+    new_dedups: Dict[bytes32, DedupCoinSpend] = {}
+    # See if this item has coin spends that are eligible for deduplication
+    assert item.npc_result.conds is not None
+    eligible_coin_ids = [bytes32(s.coin_id) for s in item.npc_result.conds.spends if s.flags & ELIGIBLE_FOR_DEDUP]
+    for coin_id in eligible_coin_ids:
+        # Pick the first coin spend with coin_id
+        # npc_result.conds.spends should be consistent with
+        # spend_bundle.coin_spends at this point, guaranteeing the presence of
+        # this coin spend in this bundle as a result
+        coin_spend_in_bundle = next(cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == coin_id)
+        solution_in_bundle = coin_spend_in_bundle.solution
+        # See if we processed an item with this coin before
+        if coin_id in dedup_coin_spends:
+            # See if the solution was identical
+            dedup_coin_spend = dedup_coin_spends[coin_id]
+            processed_solution = dedup_coin_spend.solution
+            duplicate_cost = dedup_coin_spend.cost
+            duplicate_additions = dedup_coin_spend.additions
+            if processed_solution == solution_in_bundle:
+                # Let's calculate the saved cost if we never did that before
+                if duplicate_cost is None:
+                    # If we never ran this before, additions should be empty
+                    assert len(duplicate_additions) == 0
+                    result = run_for_cost_and_additions(
+                        coin_id, coin_spend_in_bundle.puzzle_reveal, coin_spend_in_bundle.solution, item.cost
+                    )
+                    if result is None:
+                        # We failed to run this puzzle, skip this whole item
+                        consider_this_item = False
+                        cost_saving = uint64(0)
+                        break
```

Comment:
it would be easier to read (and fewer lines of code) to just say `return None` here. There's no need to assign values to those variables and break. The effect is that we'll return `None`